granulate_utils/linux/cgroups/cgroup.py

```diff
@@ -73,28 +95,230 @@ def find_v2_hierarchy(resolve_host_root_links: bool = True) -> Optional[str]:
+        interface_path = self.cgroup_full_path / interface_name
+        interface_path.write_text(data)
+
+    @abstractmethod
+    def get_subcgroup(self, controller: ControllerType, cgroup_name: str) -> CgroupCore:
+        """
+        Return a CGroup which has the given name.
+        If current cgroup is of a different name, create a subcgroup with the given name.
+        """
+        pass
+
+    @property
+    def is_v1(self) -> bool:
+        return False
+
+    @property
+    def is_v2(self) -> bool:
+        return False
+
+    @classmethod
+    def convert_outer_value_to_inner(cls, val: int) -> str:
+        return str(val)
+
+    @classmethod
+    def convert_inner_value_to_outer(cls, val: str) -> int:
+        return int(val)
+
+
+class CgroupCoreV1(CgroupCore):
+    def get_subcgroup(self, controller: ControllerType, cgroup_name: str) -> CgroupCore:
+        assert controller in CONTROLLERS
+
+        if self.cgroup_full_path.name == cgroup_name:
+            return self
+
+        return CgroupCoreV1(self._create_subcgroup(cgroup_name), self.cgroup_mount_path)
+
+    @property
+    def is_v1(self) -> bool:
+        return True
+
+
+CGROUP_V2_SUPPORTED_CONTROLLERS = "cgroup.controllers"
+CGROUP_V2_DELEGATED_CONTROLLERS = "cgroup.subtree_control"
+
+
+class CgroupCoreV2(CgroupCore):
+    def is_controller_supported(self, controller: ControllerType):
+        return controller in self.read_from_interface_file(CGROUP_V2_SUPPORTED_CONTROLLERS).split()
+
+    def _delegate_controller(self, controller: ControllerType) -> None:
+        if self.is_controller_delegated(controller):
+            return
+
+        assert self.is_controller_supported(
+            controller
+        ), f"Controller '{controller}' is not supported under {self.cgroup_full_path}"
+        self.write_to_interface_file(CGROUP_V2_DELEGATED_CONTROLLERS, f"+{controller}")
+
+    def is_controller_delegated(self, controller: ControllerType):
+        return controller in self.read_from_interface_file(CGROUP_V2_DELEGATED_CONTROLLERS).split()
+
+    def _get_parent_cgroup_for_controller(self, controller: ControllerType) -> CgroupCoreV2:
+        if self.cgroup_full_path != self.cgroup_mount_path:
+            parent_cgroup = CgroupCoreV2(self.cgroup_full_path.parent, self.cgroup_mount_path)
+            # If controller can't be delegated - fallback to root cGroup
+            if parent_cgroup.is_controller_supported(controller):
+                return parent_cgroup
+        return CgroupCoreV2(self.cgroup_mount_path, self.cgroup_mount_path)
+
+    def get_subcgroup(self, controller: ControllerType, cgroup_name: str) -> CgroupCore:
+        """
+        Create a subcgroup. There are 2 cases:
+        1. Current cGroup is the root cGroup - in this case, we create our cGroup under current cGroup.
+        2. Current cGroup isn't the root cGroup - in this case, we create a cGroup under the parent of the
+           current cGroup.
```

Comment:
Then it's not a "sub-cgroup", is it? It would be a sibling. I think having the same function for both v1 and v2 makes this code harder to manage and follow. It should be clear that the logic we apply when in v1 is clearly creating a child cgroup, and when in v2 it should be clear that we create a sibling cgroup. Let's not hide it under one function with a vague name.

Reply:
I agree with you, but this will kind of ruin the abstraction for the controllers, which will make the code much harder to use don't you think?